### PR TITLE
[Merged by Bors] - chore: bump to nightly-2023-07-01

### PR DESCRIPTION
**Pull request by kim-em**
- [x] depends on: https://github.com/leanprover/std4/pull/163
- [x] depends on: #5584
- [x] depends on: #5715
- [x] depends on: #5729
- [x] depends on: https://github.com/leanprover/std4/pull/173

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover/std4#163~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
Changes in Lean4 have caused a regression that I think we should fix upstream first, so this bump is on hold for now.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover/std4#163~~
* ~~leanprover-community/mathlib4#5584~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover/std4#163~~
* ~~leanprover-community/mathlib4#5584~~
* ~~leanprover-community/mathlib4#5715~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover/std4#163~~
* ~~leanprover-community/mathlib4#5584~~
* ~~leanprover-community/mathlib4#5715~~
* ~~leanprover-community/mathlib4#5729~~
* ~~leanprover/std4#173~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by Komyyy** [comment]
This bump makes `Nat.cast_ofNat` unavailable as a `simp` lemma.
`Nat.cast_ofNat` is used as a simp lemma by `norm_cast` to lift a numeral so this is a fatal regression.

**Comment by Komyyy** [comment]
Sorry, `Nat.cast_ofNat` is available as a `simp` lemma. It was my misunderstanding.
However, it is true that there is a regression that `norm_num` can't lift numerals.

**Comment by Komyyy** [comment]
Sorry to bother you over and over. `Nat.cast_ofNat` is often unavailable as a `simp` lemma.
It is neither absolutely unavailable nor simply available.

**Comment by Komyyy** [comment]
The cause of this problem is that a `DiscrTree.Key` which is used to match to `@OfNat.ofNat Nat 2 _` is changed from `.const OfNat.ofNat 3` to `.natLit (.natVal 2)`.
This would be fixed by registering `Nat.cast_ofNat` as a simp lemma with the modified `DiscrTree`.

**Comment by Komyyy** [comment]
Haha!🐭(THE SUDDEN MICKEY) I found the useful term elaborator `no_index`. This gives terms a `noindex` metadata. The `DiscrTree` of terms with a `noindex` metadata becomes `.star`. It is no longer needed to register `Nat.cast_ofNat` as a simp lemma with the modified `DiscrTree`.

**Comment by Komyyy** [comment]
If recent commits doesn't conflict with this branch, it will pass the CI.
This bump PR was quite a heavy job, I nearly went mad.


By the way, look! a lot of Goofy are over there.

🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!
🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck! 🐶<ah-hyuck!

File: Mathlib/AlgebraicTopology/MooreComplex.lean [hunk 1]
```diff
@@ -153,8 +162,8 @@ which maps each of these intersections of kernels to the next.
 def normalizedMooreComplex : SimplicialObject C ⥤ ChainComplex C ℕ where
   obj := obj
   map f := map f
-  map_id X := by ext (_ | _) <;> aesop_cat
-  map_comp f g := by ext (_ | _) <;> apply Subobject.eq_of_comp_arrow_eq <;> aesop_cat
+  map_id X := by ext (_ | _) <;> dsimp <;> aesop_cat
+  map_comp f g := by ext (_ | _) <;> apply Subobject.eq_of_comp_arrow_eq <;> dsimp <;> aesop_cat
```
Comment (semorrison) [inline Mathlib/AlgebraicTopology/MooreComplex.lean:166]: Could we have a note here? I'm confused why `aesop_cat` can't do the `dsimp` step.

File: Mathlib/AlgebraicTopology/MooreComplex.lean [hunk 2]
```diff
@@ -60,13 +60,22 @@ variable (X : SimplicialObject C)
...
+theorem objX_zero : objX X 0 = ⊤ :=
+  rfl
+
+theorem objX_add_one (n) :
+    objX X (n + 1) = Finset.univ.inf fun k : Fin (n + 1) => kernelSubobject (X.δ k.succ) :=
+  rfl
+
+attribute [eqns objX_zero objX_add_one] objX
+attribute [simp] objX
+
```
Comment (semorrison) [inline Mathlib/AlgebraicTopology/MooreComplex.lean:78]: Why are changes like this necessary?

File: Mathlib/AlgebraicTopology/SplitSimplicialObject.lean [hunk 3]
```diff
@@ -258,11 +258,18 @@ attribute [instance] Splitting.map_isIso
...
 #align simplicial_object.splitting.iso SimplicialObject.Splitting.iso
 
+@[simp]
+theorem iso_hom (Δ : SimplexCategoryᵒᵖ) : (iso s Δ).hom = Splitting.map X s.ι Δ :=
+  rfl
+
+@[simp]
+theorem iso_inv (Δ : SimplexCategoryᵒᵖ) : (iso s Δ).inv = inv (Splitting.map X s.ι Δ) :=
+  rfl
+
```
Comment (semorrison) [inline Mathlib/AlgebraicTopology/SplitSimplicialObject.lean:272]: Why this change? This seems like it should work.

File: Mathlib/AlgebraicTopology/MooreComplex.lean (hunk 2 above)
Comment (Komyyy) [reply Mathlib/AlgebraicTopology/MooreComplex.lean:78]: Because `objX` is unfolded to the `match` expression now.

File: Mathlib/AlgebraicTopology/SplitSimplicialObject.lean (hunk 3 above)
Comment (Komyyy) [reply Mathlib/AlgebraicTopology/SplitSimplicialObject.lean:272]: Because `Splitting.map` is unfolded in `simps` lemmas.
I wonder why `Splitting.map` isn't unfolded in the old Lean.

Comment (semorrison) [reply Mathlib/AlgebraicTopology/SplitSimplicialObject.lean:272]: Okay, I added a comment to that effect.

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge
